src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,180 @@
+end
+
+end affine_space
+
+open add_torsor affine_space
+
+section affine_subspace
+
+variables (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- An `affine_subspace k V P` is a subset of an `affine_space k V P`
+that has an affine space structure induced by a corresponding subspace
+of the `vector_space k V`. -/
+structure affine_subspace :=
+(carrier : set P)
+(direction : subspace k V)
+(nonempty : carrier.nonempty)
+(add : ∀ (p : P) (v : V), p ∈ carrier → v ∈ direction.carrier → v +ᵥ p ∈ carrier)
+(sub : ∀ (p1 p2 : P), p1 ∈ carrier → p2 ∈ carrier → p1 -ᵥ p2 ∈ direction.carrier)
+
+/-- The whole affine space as a subspace of itself. -/
+def univ_affine_subspace : affine_subspace k V P :=
+{ carrier := set.univ,
+  direction := submodule.span k set.univ,
+  nonempty := set.nonempty_iff_univ_nonempty.1 S.nonempty,
+  add := λ p v hp hv, set.mem_univ _,
+  sub := begin
+    intros p1 p2 hp1 hp2,
+    apply set.mem_bInter,
+    intros x hx,
+    rw set.mem_set_of_eq at hx,
+    exact set.mem_of_mem_of_subset (set.mem_univ _) hx
+  end }
+
+instance : inhabited (affine_subspace k V P) := ⟨univ_affine_subspace k V P⟩
+
+/-- The affine span of a nonempty set of points is the smallest affine
+subspace containing those points. (Actually defined here in terms of
+spans in vector spaces.) -/
+def affine_span (s : set P) (h : s.nonempty) : affine_subspace k V P :=
+{ carrier := span_points k V s,
+  direction := vector_span k V s,
+  nonempty := span_points_nonempty_of_nonempty k V h,
+  add := λ p v hp hv, vadd_mem_span_points_of_mem_span_points_of_mem_vector_span k V hp hv,
+  sub := λ p1 p2 hp1 hp2, vsub_mem_vector_span_of_mem_span_points_of_mem_span_points k V hp1 hp2 }
+
+end affine_subspace
+
+section affine_map
+
+variables  (k : Type*) (V1 : Type*) (P1 : Type*) (V2 : Type*) (P2 : Type*)
+    (V3 : Type*) (P3 : Type*) [field k]
+    [add_comm_group V1] [vector_space k V1] [affine_space k V1 P1]
+    [add_comm_group V2] [vector_space k V2] [affine_space k V2 P2]
+    [add_comm_group V3] [vector_space k V3] [affine_space k V3 P3]
+
+/-- An `affine_map k V1 P1 V2 P2` is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure affine_map :=
+(to_fun : P1 → P2)
+(linear : linear_map k V1 V2)
+(add : ∀ (p : P1) (v : V1), to_fun (v +ᵥ p) =  linear.to_fun v +ᵥ to_fun p)
+
+/-- Identity map as an affine map. -/
+def affine_map.id : affine_map k V1 P1 V1 P1 :=
+{ to_fun := id,
+  linear := linear_map.id,
+  add := λ p v, rfl }
+
+instance : inhabited (affine_map k V1 P1 V1 P1) := ⟨affine_map.id k V1 P1⟩
+
+/-- Composition of affine maps. -/
+def affine_map.comp (f : affine_map k V2 P2 V3 P3) (g : affine_map k V1 P1 V2 P2)
+  : affine_map k V1 P1 V3 P3 :=
+{ to_fun := f.to_fun ∘ g.to_fun,
+  linear := f.linear.comp g.linear,
+  add := begin
+    intros p v,
+    rw [function.comp_app, g.add, f.add],
+    refl
+  end
+}
+
```

Comment:
It would be good to add a `coe_to_fun` instance, and some "standard" simp lemmas. Take a look just after the definition of `ring_hom` of `linear_map` to get an idea.

Reply:
(Also the `}` should be on the same line as `end`.)

Reply:
Added that coercion and some lemmas.